homie-controller/src/lib.rs

```diff
@@ -511,23 +518,43 @@ impl HomieController {
                     && !node_id.starts_with('$')
                     && !property_id.starts_with('$') =>
             {
-                // TODO: What about values of properties we don't yet know about? They may arrive
-                // before the $properties of the node, because the "homie/node_id/+" subscription
-                // matches both.
-                let property = get_mut_property_for(
+                match get_mut_property_for(
                     devices,
                     "Got property value for",
                     device_id,
                     node_id,
                     property_id,
-                )?;
-                property.value = Some(payload.to_owned());
-                Some(Event::property_value(
-                    device_id,
-                    node_id,
-                    property,
-                    !publish.retain,
-                ))
+                ) {
+                    Ok(mut property) => {
+                        property.value = Some(payload.to_owned());
+                        Some(Event::property_value(
+                            device_id,
+                            node_id,
+                            property,
+                            !publish.retain,
+                        ))
+                    }
+
+                    Err(_) if publish.retain => {
+                        // temporarily store payloads for unknown properties to prevent
+                        // a race condition when the broker sends out the property
+                        // payloads before $properties
+
+                        early_property_values.insert(subtopic.to_owned(), payload.to_owned());
+
+                        // XXX is this return value correct? should we wait for $properties?
+                        // if so, how to return many PropertyValueChanged from there?
+                        Some(Event::PropertyValueChanged {
```

Comment:
Good question. I might leave that one for Andrew ;-)

Reply:
You could argue either way, but I'd be inclined to wait until the `$properties` come through. The match expression could be changed to return `Vec<Event>` rather than `Option<Event>`.

Reply:
This would require changing public API for `poll()`.

Reply:
Fine by me.

Reply:
Done. I have to say I still somewhat prefer #228 over this approach, though.